File: Apps/HIL/android/app/src/main/java/com/dweekly/cyrinxhil/AcousticNoiseScanner.kt
package com.dweekly.cyrinxhil

import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlin.math.cos
import kotlin.math.PI

/**
 * Background spectrum scanner for Android. Computes local background noise PSD floors,
 * identifies persistent peaks, and produces the matching 14-byte subcarrier notch mask.
 */
class AcousticNoiseScanner {
    companion object {
        /**
         * Ambiently scans a noise sample buffer, computes the Power Spectral Density (PSD),
         * runs a 51-point moving median noise-floor estimation, and yields a 14-byte subcarrier notch mask.
         */
        fun generateNotchMask(
            samples: FloatArray,
            sampleRateHz: Int,
            fftSize: Int = 1024,
            bandStartHz: Float = 18500f,
            bandEndHz: Float = 23500f,
            thresholdDB: Float = 8.0f
        ): ByteArray {
            val notchMask = ByteArray(14) { 0xFF.toByte() }
            if (samples.size < fftSize) return notchMask

            val fft = FFT(fftSize)
            val real = FloatArray(fftSize)
            val imag = FloatArray(fftSize)

            // 1. Apply Hanning window to avoid spectral leakage
            for (i in 0 until fftSize) {
                val windowVal = 0.5f * (1.0f - cos(2.0f * PI.toFloat() * i.toFloat() / (fftSize - 1).toFloat()))
                real[i] = samples[i] * windowVal
                imag[i] = 0f
            }

            // 2. Perform forward FFT
            fft.transform(real, imag, forward = true)

            // 3. Compute PSD in dB
            val psdDB = FloatArray(fftSize / 2)
            for (i in 0 until fftSize / 2) {
                val power = (real[i] * real[i]) + (imag[i] * imag[i])
                psdDB[i] = 10.0f * log10(power + 1e-12f)
            }

            // 4. Compute moving 51-point median estimation of the noise floor
            val kernelSize = 51
            val noiseFloorEst = FloatArray(psdDB.size)
            for (i in 0 until psdDB.size) {
                val startIdx = max(0, i - kernelSize / 2)
                val endIdx = min(psdDB.size - 1, i + kernelSize / 2)
                val len = endIdx - startIdx + 1
                val slice = FloatArray(len)
                System.arraycopy(psdDB, startIdx, slice, 0, len)
                slice.sort()
                val median = if (len % 2 == 1) {
                    slice[len / 2]
                } else {
                    (slice[len / 2 - 1] + slice[len / 2]) / 2.0f
                }
                noiseFloorEst[i] = median
            }

            // 5. Map to Cyrinx active subcarriers
            val binWidth = sampleRateHz.toFloat() / fftSize.toFloat()
            val startBin = max(1, ceil(bandStartHz / binWidth).toInt())
            val endBin = min((fftSize / 2) - 1, floor(bandEndHz / binWidth).toInt())
            if (startBin > endBin) return notchMask

            val activeCarrierCount = endBin - startBin + 1
            if (activeCarrierCount <= 0) return notchMask

            // Zero out notch mask before setting active carrier bits
            notchMask.fill(0)

            for (i in 0 until activeCarrierCount) {
                val bin = startBin + i
                val diff = psdDB[bin] - noiseFloorEst[bin]

                val byteIdx = i / 8
                val bitIdx = i % 8
                if (byteIdx < 14) {
                    if (diff <= thresholdDB) {
                        notchMask[byteIdx] = (notchMask[byteIdx].toInt() or (1 shl (7 - bitIdx))).toByte()
                    }
                }
            }

            return notchMask
        }
    }
}

File: Sources/Cyrinx/AcousticNoiseScanner.swift
import Foundation

#if canImport(Accelerate)
    import Accelerate
#endif

/// Background spectrum scanner that ambiently analyzes noise floors and notches out corrupted bins.
public final class AcousticNoiseScanner {
    /// Ambiently scans a noise sample buffer, computes the Power Spectral Density (PSD),
    /// runs a 51-point moving median noise-floor estimation, and yields a 14-byte subcarrier notch mask.
    public static func generateNotchMask(
        samples: [Float],
        sampleRateHz: UInt32,
        fftSize: Int = 1024,
        bandStartHz: Float = 18_500,
        bandEndHz: Float = 23_500,
        thresholdDB: Float = 8.0
    ) -> [UInt8] {
        var notchMask = [UInt8](repeating: 0xFF, count: 14)
        #if canImport(Accelerate)
            guard samples.count >= fftSize else { return notchMask }

            // 1. Setup Discrete Fourier Transform
            guard
                let dft = try? vDSP.DiscreteFourierTransform(
                    count: fftSize,
                    direction: .forward,
                    transformType: .complexComplex,
                    ofType: Float.self
                )
            else {
                return notchMask
            }

            // 2. Apply Hanning window to avoid spectral leakage
            var window = [Float](repeating: 0, count: fftSize)
            vDSP_hann_window(&window, vDSP_Length(fftSize), Int32(vDSP_HANN_NORM))

            var windowedSamples = [Float](repeating: 0, count: fftSize)
            vDSP.multiply(Array(samples.prefix(fftSize)), window, result: &windowedSamples)

            // 3. Transform to frequency domain
            var reOut = [Float](repeating: 0, count: fftSize)
            var imOut = [Float](repeating: 0, count: fftSize)
            dft.transform(
                inputReal: windowedSamples,
                inputImaginary: [Float](repeating: 0, count: fftSize),
                outputReal: &reOut,
                outputImaginary: &imOut
            )

            // 4. Compute PSD power values in decibels
            var psdDB = [Float](repeating: 0, count: fftSize / 2)
            for i in 0..<(fftSize / 2) {
                let power = (reOut[i] * reOut[i]) + (imOut[i] * imOut[i])
                psdDB[i] = 10.0 * log10(power + 1e-12)
            }

            // 5. Compute moving 51-point median to extract baseline ambient floor
            let kernelSize = 51
            var noiseFloorEst = [Float](repeating: 0, count: psdDB.count)
            for i in 0..<psdDB.count {
                let startIdx = max(0, i - kernelSize / 2)
                let endIdx = min(psdDB.count - 1, i + kernelSize / 2)
                var slice = Array(psdDB[startIdx...endIdx])
                slice.sort()
                let median: Float
                if slice.count % 2 == 1 {
                    median = slice[slice.count / 2]
                } else {
                    median = (slice[slice.count / 2 - 1] + slice[slice.count / 2]) / 2.0
                }
                noiseFloorEst[i] = median
            }

            // 6. Map background spikes to active subcarriers
            let binWidth = Float(sampleRateHz) / Float(fftSize)
            let startBin = max(1, Int(ceil(bandStartHz / binWidth)))
            let endBin = min((fftSize / 2) - 1, Int(floor(bandEndHz / binWidth)))
            guard startBin <= endBin else { return notchMask }

            let activeCarrierCount = endBin - startBin + 1
            guard activeCarrierCount > 0 else { return notchMask }

            // Clear notch mask to all zeros (all carriers default to notched/disabled)
            notchMask = [UInt8](repeating: 0, count: 14)

            // Evaluate each subcarrier's signal-to-ambient-noise level
            for i in 0..<activeCarrierCount {
                let bin = startBin + i
                let diff = psdDB[bin] - noiseFloorEst[bin]

                let byteIdx = i / 8
                let bitIdx = i % 8
                if byteIdx < 14 {
                    // If PSD power does not exceed the local floor by threshold, keep as active (1)
                    if diff <= thresholdDB {
                        notchMask[byteIdx] |= (1 << (7 - bitIdx))
                    }
                }
            }
        #endif
        return notchMask
    }
}
